packages/consent/consent-wrapper-onetrust/src/domain/wrapper.ts

```diff
@@ -0,0 +1,80 @@
+import {
+  AnyAnalytics,
+  Categories,
+  createWrapper,
+  CreateWrapperSettings,
+  resolveWhen,
+} from '@segment/analytics-consent-tools'
+
+import {
+  getConsentedGroupIds,
+  getGroupData,
+  getOneTrustGlobal,
+} from '../lib/onetrust-api'
+
+interface OneTrustOptions {
+  integrationCategoryMappings?: CreateWrapperSettings['integrationCategoryMappings']
+}
+
+export const oneTrust = (
+  analytics: AnyAnalytics,
+  options: OneTrustOptions = {}
+) =>
+  createWrapper({
+    shouldLoad: async (ctx) => {
+      await resolveWhen(() => {
+        const oneTrustGlobal = getOneTrustGlobal()
+        return (
+          oneTrustGlobal !== undefined &&
+          Boolean(getConsentedGroupIds().length) &&
+          oneTrustGlobal.IsAlertBoxClosed()
+        )
+      }, 500)
+
+      // TODO: is this the correct logic for 'user rejected all'?
+      const ids = getConsentedGroupIds()
+      const allRejected = ids.length === 1 && ids[0] === 'C0001'
+      if (allRejected) {
+        ctx.abort({
+          loadSegmentNormally: false,
+        })
+      }
+    },
+    getCategories: (): Categories => {
```

Comment:
Since `getCategories` is called a lot would it be worth memoizing the `consentedCategories` and `deniedCategories` ? 

Maybe we just invalidate the cache with `onConsentChanged` and do the reducers?

Reply:
I wouldn't memoize or do any advanced optimizations atm -- these are inexpensive operations and they all happen asynchronously